Add Kubernetes scheduler implementation

DaoWen opened the pull request:
## Changes proposed in this PR

Add Kubernetes scheduler implementation.

## Why are we making these changes?

We want the option to migrate Waiter apps from existing Marathon clusters to Kubernetes clusters.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,483 @@
+  (when-let [newest-failure (get-in pod [:status :containerStatuses 0 :lastState :terminated])]
+    (let [service-id (:service-id live-instance)
+          instance-id (:id live-instance)
+          instance-x (pod->instance-id pod \X)
```
What is `instance-x`? Maybe a better name?

DaoWen replied:
Removed while refactoring.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,483 @@
+              ;; (-> event :message (string/starts-with? "Liveness probe failed:")) #{:never-passed-health-checks}
+              ;; For now, we just assume any SIGKILL (137) with the default "Error" reason was a livenessProbe kill.
+              killed-by-k8s? (and (= 137 (:exitCode newest-failure))
+                                  (= "Error" (:reason newest-failure)))
```
Nitpick: Use destructuring on `newest-failure`.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,483 @@
+                                              :healthy? false
+                                              :id (pod->instance-id pod (count failures))
+                                              :started-at newest-failure-start-time})]
+          (doseq [[id store] [[service-id service-id->failed-instances-transient-store]
```
This `doseq` is harder to read than invoking the `swap!` calls directly

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,483 @@
+(defn- pod->ServiceInstance
+  [pod]
+  ;; FIXME - catch exceptions and return nil
+  (let [port0 (get-in pod [:spec :containers 0 :ports 0 :containerPort])]
```
Indentation is off

DaoWen replied:
Fixed.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,483 @@
+  ;; FIXME - catch exceptions and return nil
+  (let [port0 (get-in pod [:spec :containers 0 :ports 0 :containerPort])]
+  (scheduler/make-ServiceInstance
+    {:app-name (get-in pod [:metadata :labels :app])
```
Nitpick: rename `app-name` to `service-name`

DaoWen replied:
`app-name` → `k8s-name`

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,483 @@
+  (ss/try+
+    (log/info "K8s API request:" url options)
+    (let [auth-str @k8s-api-auth-str
+          result (pc/mapply mesos-utils/http-request client url
```
Maybe rename `mesos-utils`? Or extract `http-request` to another file?

DaoWen replied:
#307

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,483 @@
+  (api-request http-client k8s-object-uri
+               :request-method :patch
+               :content-type "application/json-patch+json"
+               :body (as-json ops)))
```
Nitpick: lexicographic order

DaoWen replied:
Sorted.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,483 @@
+        instances' (+ instances instances-delta)]
+    (patch-object-replicas http-client request-url instances instances')))
+
+(defn- kill-service-instance
```
Does there need to be a wait before requesting termination and force killing?

shamsimam replied:
Do we need additional logic to handle the scenario if there is an exception thrown between any of the term, scale-down, kill statements?

DaoWen replied:
> Does there need to be a wait before requesting termination and force killing?

No, that's not needed. I'm writing up some documentation explaining this.

> Do we need additional logic to handle the scenario if there is an exception thrown between any of the term, scale-down, kill statements?

Yes, that's one of my TODO items listed at the top of this PR.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,483 @@
+  (let [spec-file-path "./specs/k8s-default-pod.edn" ; TODO - allow user to provide this via the config file
+        {:strs [backend-proto cmd cpus #_disk grace-period-secs health-check-interval-secs
+                health-check-max-consecutive-failures mem ports min-instances
+                #_restart-backoff-factor run-as-user]} service-description
```
Nitpick: delete the commented destructured variables.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,483 @@
+        common-env (scheduler/environment service-id service-description
+                                          service-id->password-fn home-path)
+        port0 8080
+        template-env (into [{:name "MESOS_DIR", :value home-path}
```
Please add a comment here about backwards compatibility.

DaoWen replied:
Done.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,483 @@
+        home-path (str "/home/" run-as-user)
+        common-env (scheduler/environment service-id service-description
+                                          service-id->password-fn home-path)
+        port0 8080
```
Should this value be configurable?

DaoWen replied:
I don't see any reason why we couldn't put this in the Waiter config file, and then default to `8080` if it's not explicitly overridden.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,483 @@
+                :run-as-user run-as-user
+                :service-id service-id
+                :ssl? (= "https" backend-proto)}
+        delayed-values {:run-as-uid (->> run-as-user (shell/sh "id" "-u")
```
Extract this into a function and maintain a cache?

DaoWen replied:
I'm just going to delete this for now. What I'd really like to add a way to register custom functions named in the config file.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,483 @@
+      (api-request http-client pod-url :request-method :delete))
+    (api-request http-client replicaset-url :request-method :delete)
+    {:result :deleted
+     :message (str "K8s deleted ReplicaSet " (:app-name service))}))
```
Nitpick: lexicographic order

DaoWen replied:
Sorted.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,483 @@
+    (let [apps (get-services api-server-url http-client)
+          _ (log/info "apps =" apps)
+          app->app+instances (partial service+instances api-server-url http-client)
+          app->instances (into {} (mapv app->app+instances apps))
```
we could use `pc/map-from-keys` and then `app->app+instances` could just be `app->instances`

DaoWen replied:
Yes, that makes this much simpler.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,483 @@
+
+  (state [_]
+    {:service-id->failed-instances @service-id->failed-instances-transient-store
+     :instance-x->failed-instances @instance-x->failed-instances-transient-store}))
```
Nitpick: lexicographic order

DaoWen replied:
Fixed.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,483 @@
+         (utils/pos-int? (:socket-timeout http-options))]}
+  (let [http-client (mesos-utils/http-client-factory http-options)]
+    (when authentication
+      (let [refresh-fn (-> authentication :refresh-fn utils/resolve-symbol deref)
```
Nitpick: create a separate function for this `let` block.

DaoWen replied:
Done.

DaoWen commented on waiter/bin/ci/run-integration-tests-k8s-scheduler.sh:
```diff
@@ -0,0 +1,42 @@
+${DIR}/minikube-setup.sh
+
+# Pre-fetch our test image
+docker pull daowen/waiter-kitchen:latest
```
I don't understand your question. Are you talking about the docker image namespace? Or are you talking about how we can pre-fetch container images internally at 2σ?

shamsimam replied:
I am talking about the docket image namespace, e.g. `two-sigma/waiter-kitchen:latest`

DaoWen replied:
The namespace indicates the DockerHub account used when pushing the image. If we want to use a different DockerHub account, we can do that, but I didn't think this namespace matters that much since we pull several other dependencies from non-2σ Clojars and DockerHub namespaces.

DaoWen replied:
❌

DaoWen replied:
We could build the image in Travis (just like we do with minimesos), which would ensure we have the right kitchen. Maybe I can reuse the same image. I'll check.

DaoWen commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,483 @@
+
+(defn- service-id->app-name [service-id]
+  (let [[_ app-prefix x y z] (re-find #"([^-]+)-(\w{8})(\w+)(\w{8})$" service-id)
+        app-name-max-length (- k8s-name-length-limit 6)
```
48 is arbitrary.
6 comes from the fact that k8s pod names are auto-generated by appending a 6-character suffix to the replicaset name (e.g., `-ab123`).
I can give both of these a more descriptive name.

DaoWen commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,483 @@
+  [replicaset]
+  (let [replicaset-status (:status replicaset)
+        ready (:readyReplicas replicaset-status 0)
+        requested (get-in replicaset [:spec :replicas])
```
This entry is part of the pod `:spec`, and is required, so it doesn't need a default. The entries under `:status` are optional, which is why they have default values.

DaoWen replied:
❌

DaoWen commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,483 @@
+        started (:replicas replicaset-status 0)
+        available (:availableReplicas replicaset-status 0)
+        unavailable (:unavailableReplicas replicaset-status 0)]
+    (log/info "ReplicaSet info" replicaset)
```
Oops! I added this for debugging. This should be removed.

DaoWen commented on waiter/specs/k8s-default-pod.edn:
```diff
@@ -0,0 +1,41 @@
+ :apiVersion "extensions/v1beta1"
+ :metadata {:name #waiter/param :app-name
+            :labels {:app #waiter/param :app-name
+                     :managed-by "waiter"}
```
`:managed-by` is something I made up. It would be good if we can find an example of a more idiomatic label from the k8s community that has a similar connotation.

DaoWen commented on waiter/specs/k8s-default-pod.edn:
```diff
@@ -0,0 +1,41 @@
+{:kind "ReplicaSet"
+ :apiVersion "extensions/v1beta1"
+ :metadata {:name #waiter/param :app-name
+            :labels {:app #waiter/param :app-name
```
I don't recommend changing the `:app` label because that seems to be a strong Kubernetes convention. "Service" means something specific in the Kubernetes ecosystem, so I've been trying to avoid that term when mapping into K8s.

Using `:k8s-name` rather than `:app-name` for the template-placeholder-variable here might make sense though (as you suggested below with the `service-id->app-name` function).

shamsimam replied:
Okay, `k8s-name` sounds good.

DaoWen replied:
Renamed.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,483 @@
+  (:import (org.joda.time.format DateTimeFormat)))
+
+(def service-id->failed-instances-transient-store (atom {}))
+(def instance-x->failed-instances-transient-store (atom {}))
```
What removes entries from `instance-x->failed-instances-transient-store`?

DaoWen commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,483 @@
+            [waiter.util.utils :as utils])
+  (:import (org.joda.time.format DateTimeFormat)))
+
+(def service-id->failed-instances-transient-store (atom {}))
```
There's no reason this can't be a field in the record. I declared it this way based off the fact that the `service-id->killed-instances-transient-store` in the `waiter.scheduler` module is declared the same way, and declaring this (and the other atoms below) at the module level saves me a bunch of piping. This is one of the points I was expecting to discuss during the first round of feedback.

DaoWen commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,483 @@
+
+(def service-id->failed-instances-transient-store (atom {}))
+(def instance-x->failed-instances-transient-store (atom {}))
+(def k8s-api-auth-str (atom nil))
```
See above.

DaoWen commented on waiter/src/waiter/settings.clj:
```diff
@@ -268,6 +268,11 @@
                                  :force-kill-after-ms 60000
                                  :framework-id-ttl 900000}
+                      :kubernetes {:factory-fn 'waiter.scheduler.kubernetes/kubernetes-scheduler
+                                   :http-options {:conn-timeout 10000
```
Done.

DaoWen commented on waiter/specs/k8s-default-pod.edn:
```diff
@@ -0,0 +1,41 @@
+{:kind "ReplicaSet"
```
I think it would be nice having the ability to patch things in the pod template spec (e.g., add a new environment variable, or update the image uri) without having to rebuild the JAR file. Is there a reason why we wouldn't want to treat this like another configuration file?

DaoWen replied:
The path to the pod spec file is now part of the `:kubernetes` settings in our Waiter config file.

DaoWen commented on waiter/specs/k8s-default-pod.edn:
```diff
@@ -0,0 +1,41 @@
+ :metadata {:name #waiter/param :app-name
+            :labels {:app #waiter/param :app-name
+                     :managed-by "waiter"}
+            :annotations {:waiter/app-status "live"
```
`app-name` → `k8s-name`

DaoWen commented on waiter/specs/k8s-default-pod.edn:
```diff
@@ -0,0 +1,41 @@
+                                 :managed-by "waiter"}}
+        :template {:metadata {:labels {:app #waiter/param :app-name
+                                       :managed-by "waiter"}
+                              :annotations {
```
Deleted.

DaoWen commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,483 @@
+       :task-count requested
+       :task-stats {:running ready
+                    :staged (- requested ready)
+                    :healthy available
```
Sorted.

DaoWen commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,483 @@
+        unavailable (:unavailableReplicas replicaset-status 0)]
+    (log/info "ReplicaSet info" replicaset)
+    (scheduler/make-Service
+      {:app-name (get-in replicaset [:metadata :name])
```
`app-name` → `k8s-name`

DaoWen commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,483 @@
+
+(defn- replicaset->Service
+  ;; FIXME - catch exceptions and return nil
+  [replicaset]
```
I think I prefer `replicaset`. It's always written `ReplicaSet` in the k8s documentation (never "replica set"). One can argue that PascalCase words should be kabob-case in Clojure—but I also feel that `replicaset` leads to less ambiguous names (e.g., `replicaset-delete` is clearer than `replica-set-delete`, which could be "delete the replica set" or "set the delete attribute on a replica").

DaoWen replied:
❌

DaoWen commented on waiter/specs/k8s-default-pod.edn:
```diff
@@ -0,0 +1,41 @@
+                   :spec {:containers [{:name #waiter/param :app-name
+                                        :command ["/bin/sh" "-c" #waiter/param :cmd]
+                                        :env #waiter/param :env
+                                        :image "daowen/waiter-kitchen:latest"
```
If this pod spec is being treated like a config file, then there's no need to pull the image reference from here and move it into the Waiter config; however, if we do move this spec into the resources folder (pack it into the Waiter JAR), then it might make more sense to set it in the config file. That would probably also mean we'd want to set the command in the config file, since the command and the image are often tightly coupled.

DaoWen commented:
Closing and reopening to trigger Travis

DaoWen commented:
@sradack @shamsimam - this is ready for another pass

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,605 @@
+          (<= attempt max-conflict-retries)
+          (recur (inc attempt) (get-replica-count scheduler replicaset-url)))))))
+
+(defn- kill-service-instance
```
What happens if any of the steps in the delete fail? E.g. deleting the instance succeeds but subsequent calls fail. The autoscaler will think the instance was not killed successfully and a subsequent request maybe assigned this killed instance causing the request to fail.

DaoWen replied:
You were right that this needed to be revisited. I added a patch to fix this. Now the operation returns success as long as the first `:delete` operation happens. I also updated the `pod->ServiceInstance` caller logic so that deleted pods (i.e., pods with the `deletionTimestamp` present) are no longer visible to Waiter.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,605 @@
+                :ssl? (= "https" backend-proto)}
+        edn-opts {:readers {'waiter/param params
+                            'waiter/param-str (comp str params)
+                            'waiter/port #(+ port0 %)
```
Should port also be exposed as params `PORT0`, `PORT1` and so on based on number of requested ports?

DaoWen replied:
I think it's already doing what you're requesting. E.g., in the default ReplicaSet spec:
```clojure
:ports [{:containerPort #waiter/port 0}]
```
`#waiter/port 0` above could be replaced with `#waiter/port 5` or any other valid number.

shamsimam replied:
I meant refer to it as `#waiter/param :PORT0`

DaoWen replied:
Is there an advantage to writing `#waiter/param :PORT0` vs `#waiter/port 0`? Is the former less confusing? I feel like we should pick just one or the other.

shamsimam replied:
I think I am more familiar with what `PORT0` means, whereas I had to lookup the code to understand what `#waiter/port 0` was doing.

DaoWen replied:
All of the other `param` values are lower-case. What about `#waiter/env PORT0`? I.e., we'd be exposing all of the values from the provided environment (including `PORT0` – `PORT9`, if available) for the user to reference inside the spec.

DaoWen replied:
It might also make sense to sanitize the arguments to the `param` and `env` reader functions (e.g., pass them thru `keyword` or `name`) so symbols/strings/keywords all work sensibly.

DaoWen replied:
I just thought of an issue with my `#waiter/env` idea: all environment values are *strings*, but we need an *integer* for this use case. There's no reason that `#waiter/param :port0` couldn't return an integer, but the `#waiter/env` idea doesn't really work here.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,605 @@
+                            'waiter/param-str (comp str params)
+                            'waiter/port #(+ port0 %)
+                            'waiter.fn/into (fn [[xs ys]] (into xs ys))
+                            'waiter.fn/lower-case string/upper-case
```
Should this be `string/lower-case`?

DaoWen replied:
Yes, that would make more sense.

DaoWen replied:
Removed along with the other `waiter.fn` entries.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,605 @@
+        edn-opts {:readers {'waiter/param params
+                            'waiter/param-str (comp str params)
+                            'waiter/port #(+ port0 %)
+                            'waiter.fn/into (fn [[xs ys]] (into xs ys))
```
Are these `waiter.fn`s used? If not, we should remove them.

DaoWen replied:
Done.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,605 @@
+
+(defn- create-service
+  "Reify a Waiter Service as a Kuberentes ReplicaSet."
+  [service-id descriptor {:keys [api-server-url http-client] :as scheduler} service-id->password-fn]
```
Nitpick: can destructure `service-description` in the `descriptor` here.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,605 @@
+                         "/apis/extensions/v1beta1/namespaces/"
+                         (service-description->namespace service-description)
+                         "/replicasets")
+        response-json (api-request http-client request-url
```
Nitpick: we can destructure `kind` here

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,605 @@
+                                   :request-method :post)]
+    (when-not (= "ReplicaSet" (:kind response-json))
+      (log/error "Invalid response from ReplicaSet create request"
+                 (as-json response-json)))
```
Why convert the response back to a json string in the log?

DaoWen replied:
I can't remember why I did that... Deleted the `as-json` call.

DaoWen replied:
Actually, I don't know why I added this check at all. Kubernetes would never return a different kind of object when we're posting to the repicasets endpoint—and if it did, we'd want to abort, not log the error and continue... I'm deleting this whole check.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,605 @@
+   Ensures that all controlled Pods are deleted before removing the ReplicaSet.
+   This operation will leave Waiter in a sane state on a failure."
+  [{:keys [api-server-url http-client] :as scheduler} service]
+  (when-not service
```
Why can this happen?

DaoWen replied:
The `service-id->service` call that feeds into this can return `nil` if it fails (e.g., 404).

shamsimam replied:
Should the delete call be avoided when that happens?

DaoWen replied:
Done. Now `service-id->service` throws on 404 errors.

shamsimam commented on waiter/specs/k8s-default-pod.edn:
```diff
@@ -0,0 +1,38 @@
+{:kind "ReplicaSet"
+ :apiVersion "extensions/v1beta1"
```
Should this come from the Clojure code as many of the api queries use this version string too? Maybe it needs to be a constant in the kubernetes scheduler code?

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,605 @@
+                       [{:op :replace :path "/metadata/annotations/waiter~1app-status" :value "killed"}
+                        {:op :replace :path "/spec/replicas" :value 0}])
+    (doseq [pod (get-replicaset-pods scheduler service)
+            :let [pod-url (->> pod :metadata :selfLink (str api-server-url))]]
```
Nitpick: I would have preferred the `let` block inside the `doseq` body.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,605 @@
+    (patch-object-json replicaset-url http-client
+                       [{:op :replace :path "/metadata/annotations/waiter~1app-status" :value "killed"}
+                        {:op :replace :path "/spec/replicas" :value 0}])
+    (doseq [pod (get-replicaset-pods scheduler service)
```
Is the presence of these pods an issue if there is a concurrent `create-app` call while the delete is in progress?

shamsimam replied:
I missed that setting the replicas to zero would still mean that the app exists and hence would cause the `create-app` call to become a no-op.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,605 @@
+    {:message (str "Kubernetes deleted ReplicaSet for " (:k8s-name service))
+     :result :deleted}))
+
+(defn service-id->service
```
Looks like most places this is invoked we are mainly interested in the metadata associated with the service (`id`, `namespace`, `pod-name` etc). For those call sites, we can then have a separate function and rely on a cached version of this information of a new data structure and pass it around? This will avoid having to make the k8s api call over the wire to build the metadata.

DaoWen replied:
I agree, but I'll leave that optimization for a future PR. We'll be adding local caching of state when I add a watcher to get pushed state updates from the Kubernetes master, and I think this would follow naturally from that refactoring.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,605 @@
+  (get-apps->instances [this]
+    (->> this
+         get-services
+         (pc/map-from-keys (partial instances-breakdown this))))
```
Nitpick: I find this more readable:
`(pc/map-from-keys #(instances-breakdown this %) (get-services this))`

DaoWen replied:
Done.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,605 @@
+(defn- start-auth-renewer
+  "Initialize the k8s-api-auth-str atom,
+   and optionally start a chime to periodically referesh the value."
+  [{:keys [refresh-delay-minutes refresh-fn]}]
```
Nitpick: we can include a precondition check for `refresh-fn` not being nil.

DaoWen replied:
Added preconditions:
```clojure
  {:pre [(utils/pos-int? refresh-delay-minutes)
         (fn? refresh-fn)]}
```

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,605 @@
+                           (reset! k8s-api-auth-str auth-str')))]
+    (auth-update-fn)
+    (when [refresh-delay-minutes]
+      (assert (utils/pos-int? refresh-delay-minutes))
```
We can move this assertion to a precondition check.

DaoWen replied:
Done.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,605 @@
+      (assert (utils/pos-int? refresh-delay-minutes))
+      (-> refresh-delay-minutes
+          t/minutes
+          (du/start-timer-task auth-update-fn)))))
```
We need to set the `delay-ms` option to trigger the timer task after the delay.

DaoWen replied:
Done.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,605 @@
+         (utils/non-neg-int? max-conflict-retries)
+         (utils/pos-int? max-name-length)
+         (integer? pod-base-port)
+         (<= 1024 pod-base-port 49151)  ; registered port range
```
Can we add some extra documentation somewhere on why this is the registered port range, e.g. in the `config-full.edn`?

DaoWen replied:
If I remember correctly, Java throws if you try to bind to a port above 49151. Ports below 1024 require root, and are more likely to conflict with other system services.

If you agree that those are good reasons to restrict the port range, then I can explain that in config-full.edn. If you thin I'm being overly restrictive here, I can loosen the preconditions.

shamsimam replied:
We can follow the intersection of what Mesos agents allow as their port ranges and the ranges that work with Kubernetes.

DaoWen replied:
It looks like Mesos lets you specify custom ranges of ports that a slave can expose to containers as resources, but by default uses a small subrange of the "registered" ports: 31000 to 32000.<sup>[link](https://github.com/apache/mesos/blob/0f0909c89578dbe9ea23d5fa4d980712789a2c32/src/slave/constants.hpp#L108-L109)

shamsimam replied:
We can set it to 31000 by default.

Should we allow users to configure in any valid positive integer (1 - 65535)?

DaoWen replied:
We should probably do 65526 as the upper bound (since we allow using up to 10 ports), but both the 31000 default and the less-restrictive range sound reasonable.

DaoWen replied:
Done.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,605 @@
+         (utils/pos-int? max-name-length)
+         (integer? pod-base-port)
+         (<= 1024 pod-base-port 49151)  ; registered port range
+         (.exists (io/as-file rs-spec-file-path))
```
👍

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,605 @@
+         (integer? pod-base-port)
+         (<= 1024 pod-base-port 49151)  ; registered port range
+         (.exists (io/as-file rs-spec-file-path))
+         (some? (io/as-url url))]}
```
👍

shamsimam commented on waiter/test/waiter/scheduler/kubernetes_test.clj:
```diff
@@ -0,0 +1,921 @@
+            [clojure.walk :as walk]
+            [plumbing.core :as pc]
+            [slingshot.slingshot :as ss]
+            [waiter.scheduler.kubernetes :refer :all]
```
Nitpick: this should be below `[waiter.scheduler :as scheduler]`

DaoWen replied:
Done.

shamsimam commented on waiter/test/waiter/scheduler/kubernetes_test.clj:
```diff
@@ -0,0 +1,921 @@
+            [waiter.scheduler.kubernetes :refer :all]
+            [waiter.scheduler :as scheduler]
+            [waiter.util.date-utils :as du])
+  (:import waiter.scheduler.kubernetes.KubernetesScheduler
```
Nitpick: Can we place this below `(waiter.scheduler...)`?

shamsimam replied:
Nitpick: Can we use `(waiter.scheduler.kubernetes KubernetesScheduler)` instead?

DaoWen replied:
Done.

DaoWen commented on waiter/bin/test.sh:
```diff
@@ -10,7 +10,7 @@
 # WAITER_TEST_KITCHEN_CMD are set, and sets them to reasonable defaults if not.
 
-set -v
+set -x
```
Yes, I missed commit when I was dropping unrelated changes. Sorry!

DaoWen replied:
Dropped.

DaoWen commented on waiter/config-k8s.edn:
```diff
@@ -0,0 +1,35 @@
+
+ :scheduler-config {:kind :kubernetes
+                    :kubernetes {:url "http://localhost:8001"
+                                 :rs-spec-file-path "./specs/k8s-default-pod.edn"}}
```
`rs` is the standard short-form of `replicaset` used in the Kubernetes ecosystem. E.g., you can run `kubectl get rs` to get the same results as `kubectl get replicasets`. I was just worried about the name being too long, but I can change this.

DaoWen commented on waiter/integration/waiter/metrics_output_test.clj:
```diff
@@ -99,7 +99,7 @@
        (str ~metrics-data)))
 
-(deftest ^:parallel ^:integration-fast test-metrics-output
+(deftest ^:parallel ^:integration-slow test-metrics-output
```
Again, botched rebase. Sorry. Dropping.

DaoWen commented on waiter/integration/waiter/autoscaling_test.clj:
```diff
@@ -37,7 +37,7 @@
       (delete-service waiter-url service-id))))
 
-(deftest ^:parallel ^:integration-slow test-scaling-healthy-app
+(deftest ^:parallel ^:integration-slow ^:resource-heavy test-scaling-healthy-app
```
Ugh. I really botched that rebase. Dropping.

DaoWen commented on waiter/project.clj:
```diff
@@ -23,8 +23,10 @@
-                   :integration-slow (every-pred :integration-slow (complement :explicit))
                    :integration-fast (every-pred :integration-fast (complement :explicit))
+                   :integration-slow (every-pred :integration-slow (complement :explicit))
+                   :integration-slow-heavy (every-pred :integration-slow :resource-heavy)
```
Yes. I was thinking this was merging the `:integration-fast` and `:integration-slow` predicates—but that's not how it works. Thanks.

DaoWen replied:
Dropped.

DaoWen commented on waiter/resources/log4j.properties:
```diff
@@ -28,3 +28,13 @@ log4j.appender.RequestLogAppender.File=log/${waiter.logFilePrefix}request.log
+log4j.category.Kubernetes=DEBUG, KubernetesAppender
+log4j.additivity.Kubernetes=false
+
+log4j.appender.KubernetesAppender=org.apache.log4j.DailyRollingFileAppender
```
I like that idea. Just rename it Scheduler, right?

DaoWen replied:
I think this is very helpful for Kubernetes debugging. I also like tailing this file during demos to show how the Waiter operations we're doing turn into Kubernetes API requests. I think it's nice having it separate. Do you have strong feelings about adding another log file?

shamsimam replied:
I initially thought you only kept it for being able to debug. I'm good with having a separate file for the scheduler.

DaoWen replied:
The Kubernetes log is now a more generic Scheduler log. I also moved the logging macro to the scheduler namespace since that seemed more fitting; i.e., `(k8s-log ...)` → `(scheduler/log ...)`.

DaoWen commented on waiter/specs/k8s-default-pod.edn:
```diff
@@ -0,0 +1,38 @@
+ :apiVersion "extensions/v1beta1"
+ :metadata {:name #waiter/param :k8s-name
+            :labels {:app #waiter/param :k8s-name
+                     :managed-by "waiter"}
```
That's implied by the namespace here.

DaoWen replied:
All of the examples in the Kubernetes docs that I've seen use the `app` label for the application name. E.g.: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#more-practical-use-cases

shamsimam replied:
Do we want to add it anyway in case we choose some other namespace mechanism later?

DaoWen replied:
My impression is that adding extra labels to a Kubernetes object is pretty cheap, so I guess there's no harm in adding it.

shamsimam replied:
sgtm :)

DaoWen replied:
OK, thinking about this more, this seems very redundant for something that we don't anticipate using anytime soon. Since it's a fairly lightweight thing, it will be simple to add it in the future if we decide to stop assuming the namespace⇔run-as-user correspondence.

DaoWen commented on waiter/specs/k8s-default-pod.edn:
```diff
@@ -0,0 +1,38 @@
+{:kind "ReplicaSet"
+ :apiVersion "extensions/v1beta1"
+ :metadata {:name #waiter/param :k8s-name
+            :labels {:app #waiter/param :k8s-name
```
All of the examples in the Kubernetes docs that I've seen use the `app` label for the application name. E.g.: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#more-practical-use-cases

shamsimam replied:
okay.

DaoWen commented on waiter/specs/k8s-default-pod.edn:
```diff
@@ -0,0 +1,38 @@
+ :metadata {:name #waiter/param :k8s-name
+            :labels {:app #waiter/param :k8s-name
+                     :managed-by "waiter"}
+            :annotations {:waiter/app-status "live"
```
I think using `service-status` here would make sense since this is in the scope of waiter. I'll rename it.

DaoWen commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,605 @@
+      (comment "Returning nil on failure."))))
+
+(defn api-request
+  "Make an HTTP request to the Kuberenets API server using the configured authentication.
```
I think "an HTTP" is correct here.<sup>[link](http://blog.apastyle.org/apastyle/2012/04/using-a-or-an-with-acronyms-and-abbreviations.html)</sup> A similar example is "an MP3."

DaoWen commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,605 @@
+(defmacro k8s-log
+  "Log Kuberenets-specific messages."
+  [& args]
+  `(log/log "Kubernetes" :debug nil (print-str ~@args)))
```
Yes, that would make sense. I'll add that.

DaoWen commented on waiter/specs/k8s-default-pod.edn:
```diff
@@ -0,0 +1,38 @@
+                          :waiter/service-id #waiter/param :service-id}}
+ :spec {:replicas #waiter/param :min-instances
+        :selector {:matchLabels {:app #waiter/param :k8s-name
+                                 :managed-by "waiter"}}
```
I think this is analogous to the prefixes we can set for waiter jobs in Marathon. I'll move this to the `:kubernetes` section of the Waiter config file, and then use that value here (indirectly).

DaoWen commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,605 @@
+  "Get all Waiter Services (reified as ReplicaSets) running in this Kuberentes cluster."
+  [{:keys [api-server-url http-client] :as scheduler}]
+  (->>
+    "/apis/extensions/v1beta1/replicasets?labelSelector=managed-by=waiter"
```
I don't think so.

DaoWen commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,605 @@
+  (->> (str api-server-url
+            "/api/v1/namespaces/"
+            namespace
+            "/pods?labelSelector=app="
```
I don't think so.

DaoWen replied:
To clarify, I don't think it's required by the relevant RFCs (I spent a fair amount of time looking at those when we were doing the query string pass-thru stuff), and more importantly, it works fine both with our client and with the Kubernetes API server.

DaoWen commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,605 @@
+  "Update the replica count in the given Kubernetes object's spec."
+  [k8s-object-uri http-client replicas replicas']
+  (patch-object-json http-client k8s-object-uri
+                     ;; NOTE: ~1 is JSON-patch escape syntax for a "/" in a key name.
```
Convention.

shamsimam replied:
Wouldn't using `waiter-` prefix simplify our lives and avoid this comment?

DaoWen replied:
Maybe `waiter:`? I'll need to confirm that doesn't cause problems elsewhere.

DaoWen replied:
I switched it to `waiter-`.

DaoWen commented on waiter/project.clj:
```diff
@@ -23,8 +23,10 @@
                    :integration-fast (every-pred :integration-fast (complement :explicit))
+                   :integration-slow (every-pred :integration-slow (complement :explicit))
+                   :integration-slow-heavy (every-pred :integration-slow :resource-heavy)
+                   :integration-slow-lite (every-pred :integration-slow (complement :resource-heavy))
```
Dropped.

DaoWen commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,605 @@
+(defn- as-json
+  "Convert Clojure data structures to JSON, preserving namespaces on keys."
+  [data]
+  (json/write-str data :key-fn #(if (keyword? %) (kw->str %) %)))
```
Replacing `/` with `-` in the annotation keys eliminated the need for these helpers.

DaoWen commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,605 @@
+                 update-in [service-id] assoc newest-failure-id newest-failure-instance))))))
+
+(defn- pod->ServiceInstance
+  "Convert a Kuberentes Pod JSON response into a Waiter Service Instance record."
```
Kubernetes was misspelled in like 10 places!
`git grep 'Kub' | fgrep -v Kubernetes`
...
All fixed now!

DaoWen commented on waiter/src/waiter/settings.clj:
```diff
@@ -269,6 +269,12 @@
              ; throttles the rate at which kill requests are sent to the scheduler
              :inter-kill-request-wait-time-ms 1000}
    :scheduler-config {:kind :marathon
+                      :kubernetes {:factory-fn 'waiter.scheduler.kubernetes/kubernetes-scheduler
```
Both done.

DaoWen commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,605 @@
+   Note that a new Waiter Service Instance is created each time a pod restarts."
+  ([pod] (pod->instance-id pod (get-in pod [:status :containerStatuses 0 :restartCount])))
+  ([pod restart-count]
+   (let [pod-name (get-in pod [:metadata :name])
```
Done.

DaoWen commented:
@shamsimam - Ready for another pass. Still working on getting the integration tests to pass reliably, but that's going to involve changes from some separate PRs.

shamsimam commented on waiter/config-full.edn:
```diff
@@ -263,6 +263,45 @@
+                                                  ;; Nullary function which can be called periodically (or just once if :refresh-delay-minutes is omitted)
+                                                  ;; to obtain an Authorization string for use when authenticating with the Kubernetes API server.
+                                                  ;; The function must return a string that can be used as the value for the Authorization HTTP header.
+                                                  :refresh-fn waiter.scheduler.kubernetes/mock-auth-refresh}
```
Can we point to an actual implementation instead of having a mock implementation in the main source code?

DaoWen replied:
Changed this mock implementation to a less contrived sample:
```clojure
(defn authorization-from-environment []
  "Sample implementation of the authentication string refresh function.
   Returns a string to be used as the value for the Authorization HTTP header,
   reading the string from the WAITER_K8S_AUTH_STRING environment variable."
  (log/info "Called waiter.scheduler.kubernetes/authorization-from-environment")
  (System/getenv "WAITER_K8S_AUTH_STRING"))
```

DaoWen replied:
Done.

shamsimam commented on waiter/specs/k8s-default-pod.edn:
```diff
@@ -0,0 +1,39 @@
+ :apiVersion #waiter/param :replicaset-api-version
+ :metadata {:name #waiter/param :k8s-name
+            :labels {:app #waiter/param :k8s-name
+                     :managed-by "waiter"}
```
This value should come from the Clojure code as it is used in the queries too. Or we should have a comment here that specifies that this value should not be changed.

DaoWen replied:
Done.

shamsimam commented on waiter/specs/k8s-default-pod.edn:
```diff
@@ -0,0 +1,39 @@
+                          :waiter-service-id #waiter/param :service-id}}
+ :spec {:replicas #waiter/param :min-instances
+        :selector {:matchLabels {:app #waiter/param :k8s-name
+                                 :managed-by "waiter"}}
```
This value should come from the Clojure code as it is used in the queries too. Or we should have a comment here that specifies that this value should not be changed.

DaoWen replied:
Done.

shamsimam commented on waiter/specs/k8s-default-pod.edn:
```diff
@@ -0,0 +1,39 @@
+        :selector {:matchLabels {:app #waiter/param :k8s-name
+                                 :managed-by "waiter"}}
+        :template {:metadata {:labels {:app #waiter/param :k8s-name
+                                       :managed-by "waiter"}
```
This value should come from the Clojure code as it is used in the queries too. Or we should have a comment here that specifies that this value should not be changed.

DaoWen replied:
Done.

shamsimam commented on waiter/specs/k8s-default-pod.edn:
```diff
@@ -0,0 +1,39 @@
+                              :annotations {:waiter-port-count #waiter/param-str :port-count
+                                            :waiter-protocol #waiter/param :backend-protocol
+                                            :waiter-service-id #waiter/param :service-id}}
+                   :spec {:containers [{:name #waiter/param :k8s-name
```
Nitpick: lexicographic order

DaoWen replied:
Sorted.

shamsimam commented on waiter/src/waiter/scheduler.clj:
```diff
@@ -37,6 +37,11 @@
+(defmacro log
+  "Log Scheduler-specific messages."
+  [& args]
+  `(log/log "Scheduler" :debug nil (print-str ~@args)))
```
Query: does this include the CID in the logs?

DaoWen replied:
To clarify, I updated the logging format in the log4j properties file to include the CID.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,594 @@
+            [waiter.util.utils :as utils])
+  (:import (org.joda.time.format DateTimeFormat)))
+
+(defn mock-auth-refresh []
```
I think we should not have a mock implementation in the source directory.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,594 @@
+  (try
+    (let [port0 (get-in pod [:spec :containers 0 :ports 0 :containerPort])]
+      (scheduler/make-ServiceInstance
+        {:k8s-name (get-in pod [:metadata :labels :app])
```
We should use namespace keywords for k8s specific entries, e.g.`k8s/name`, `k8s/namespace`, etc.

DaoWen replied:
Done.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,594 @@
+        (comment "App does not exist."))))
+
+  (create-app-if-new [this service-id->password-fn descriptor]
+    (let [service-id (:service-id descriptor)]
```
We can use destructuring for this.

DaoWen replied:
Done.

shamsimam commented on waiter/test/waiter/scheduler/kubernetes_test.clj:
```diff
@@ -0,0 +1,922 @@
+     {:max-conflict-retries 5
+      :max-name-length 63
+      :replicaset-spec-file-path default-rs-spec-path
+      :pod-base-port 8080
```
nitpick: lexicographic order

DaoWen replied:
Sorted.

DaoWen commented:
@sradack - I think this is ready for your feedback.

sradack commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,606 @@
+                      (subs 0 prefix-max-length))]
+    (str app-prefix' suffix)))
+
+(pc/defnk replicaset->Service
```
Interesting use of `defnk`.  I hadn't thought about using `defnk` outside of the graph-compiled stuff.  Seems nicer than using Clojure's default destructuring in this case.

sradack commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,606 @@
+                                   :request-method :post)]
+    (replicaset->Service response-json)))
+
+(defn- delete-service
```
How are we cleaning things up if this fails after we marked the app as `killed`?  Do we ignore `killed` apps somewhere?

DaoWen replied:
I changed the implementation to leverage K8s GC. The `delete-service` implementation is now much simpler, and the `waiter-app-status` attribute has been removed.

sradack commented on waiter/config-full.edn:
```diff
@@ -263,6 +263,49 @@
+                                 ;; Port number used to set the $PORT0 (service port) for Waiter service instances.
+                                 ;; If more than one port is requested, $PORT1 thru $PORT9 will be defined sequentially after $PORT0
+                                 ;; (e.g., if pod-base-port is 8080, then $PORT0==8080 and $PORT9==8089):
+                                 :pod-base-port 31000
```
Ports should be "random"ized.  I'm thinking something based upon the `service-id`, confined within a certain range.

sradack commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,606 @@
+;; Kubernetes Pods have a unique 5-character alphanumeric suffix preceded by a hyphen.
+(def ^:const pod-unique-suffix-length 5)
+
+(defn- service-id->k8s-app-name [{:keys [max-name-length] :as scheduler} service-id]
```
It would be nice to have some examples.

shamsimam replied:
Should we consider using only the hash part from the service-id and name it `service-hash`?

sradack commented on .travis.yml:
```diff
@@ -38,6 +38,16 @@ matrix:
+    - env: "NAME='Waiter integration tests: kubernetes-fast'"
+      services: docker
+      before_script: cd waiter && voom-build
+      script: ./bin/ci/run-integration-tests-k8s-scheduler.sh parallel-test integration-fast
```
Where are the integration tests in this PR?

sradack commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,606 @@
+          (recur (inc attempt) (get-replica-count scheduler replicaset-url)))))))
+
+(defn- kill-service-instance
+  "Safely kill the Kubernetes pod corresponding to the given Waiter Service Instance.
```
It would be great to discuss why this method of killing is safe.

DaoWen commented:
The commit history was getting ridiculously long here, so I squashed them. This build is green in a PR on my private fork:

https://github.com/DaoWen/waiter/pull/5

However, that branch has some extra (unrelated) commits that aren't in our master branch yet. The build on this PR won't be green until after we've merged the following changes:

1. Python Kitchen (#374)
2. Integration heavy/lite lein test selectors (#385)

DaoWen commented on waiter/resources/log4j.properties:
```diff
@@ -28,3 +28,13 @@ log4j.appender.RequestLogAppender.File=log/${waiter.logFilePrefix}request.log
 log4j.appender.RequestLogAppender.layout.ConversionPattern=%m%n
+
+log4j.category.Scheduler=DEBUG, SchedulerAppender
+log4j.additivity.Scheduler=false
```
As per our conversation today, I updated the log4j config so the Kubernetes API calls get logged to the main waiter log as well as the separate scheduler log. Errors will also be logged to the waiter-error log.
```
log4j.category.Scheduler=DEBUG, InfoAppender, ErrorAppender, SchedulerAppender
```

DaoWen commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,606 @@
+  "Sample implementation of the authentication string refresh function.
+   Returns a string to be used as the value for the Authorization HTTP header,
+   reading the string from the WAITER_K8S_AUTH_STRING environment variable."
+  (log/info "Called waiter.scheduler.kubernetes/authorization-from-environment")
```
All log messages are now lower-case.

DaoWen commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,606 @@
+
+(defn- killed-by-k8s?
+  "Determine whether a pod was killed (restarted) by its corresponding Kubernetes liveness checks."
+  [pod-terminated-info]
```
Done.

DaoWen commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,606 @@
+                                             {:flags failure-flags
+                                              :healthy? false
+                                              :id newest-failure-id
+                                              :started-at newest-failure-start-time})]
```
Done.

DaoWen commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,606 @@
+                         (get-in [:status :startTime])
+                         (timestamp-str->datetime))}))
+    (catch Throwable e
+      (log/error e "Error converting pod to waiter service instance" pod)
```
All log messages are now lower-case.

DaoWen commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,606 @@
+   If data is provided via :body, the application/json content type is added automatically.
+   The response payload (if any) is automatically parsed to JSON."
+  [client url & {:keys [body content-type request-method] :as options}]
+  (scheduler/log "Making request to K8s API server:" url request-method body)
```
All log messages are now lower-case.

DaoWen commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,606 @@
+                            (cond-> options
+                              auth-str (assoc-in [:headers "Authorization"] auth-str)
+                              (and (not content-type ) body) (assoc :content-type "application/json")))]
+      (scheduler/log "Response from K8s API server:" (json/write-str result))
```
All log messages are now lower-case.

DaoWen commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,606 @@
+      (scheduler/log "Response from K8s API server:" (json/write-str result))
+      result)
+    (catch [:status 400] _
+      (log/error "Malformed K8s API request: " url options))
```
All log messages are now lower-case.

DaoWen commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,606 @@
+    (catch [:status 400] _
+      (log/error "Malformed K8s API request: " url options))
+    (catch [:client client] response
+      (log/error "Request to K8s API server failed: " url options body response)
```
All log messages are now lower-case.

shamsimam commented:
>  Fixed all of the nitpicks.

Sweet, thanks

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,632 @@
+;; Kubernetes Pods have a unique 5-character alphanumeric suffix preceded by a hyphen.
+(def ^:const pod-unique-suffix-length 5)
+
+(defn service-id->k8s-app-name [{:keys [max-name-length] :as scheduler} service-id]
```
Should we consider using only the hash part from the `service-id` as the `k8s-app-name`; we could expose it as the `service-hash` in the descriptor?

DaoWen replied:
If we want to optimize for the lowest possible collision probabilities, that seems like a good solution. Right now I'm trying to strike a balance between keeping the output of `kubectl get pods` readable, and keeping the probability of name conflicts in k8s low.

I'd like to keep this as-is for the current PR, but I'm open to revisiting this discussion later.

DaoWen commented:
@sradack & @shamsimam - #374 and #385 are finally both merged, and we've got a green build! Unless you have some additional feedback that you want included in this patch, then I think this is *finally* RTM! 🎉

shamsimam commented:
We should updated our documents that refer exclusively to Mesos as part of this PR! The https://github.com/twosigma/waiter/blob/master/README.md , https://github.com/twosigma/waiter/blob/master/waiter/README.md , https://github.com/twosigma/waiter/blob/master/waiter/docs/service-description.md files , among others, need to change.

DaoWen commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,646 @@
+  "Get all Waiter Service Instances associated with the given Waiter Service."
+  [{:keys [api-server-url http-client] :as scheduler} basic-service-info]
+  (vec (for [pod (get-replicaset-pods scheduler basic-service-info)
+             :when (live-pod? pod)]
```
`:when` is equivalent to `filter`. Maybe you're confusing `:when` and `:while`?

shamsimam replied:
Indeed I was :(

sradack commented on waiter/specs/k8s-default-pod.edn:
```diff
@@ -0,0 +1,39 @@
+{:kind "ReplicaSet"
```
Let's move this into the code.  It will simplify customizations to the `ReplicaSet`.

shamsimam commented on waiter/config-full.edn:
```diff
@@ -288,21 +288,15 @@
+                                                           ;; any additional configuration data needed for the implementation should be included here.
+                                                           :factory-fn waiter.scheduler.kubernetes/default-replicaset-builder
+                                                           ;; The default factory function accepts an option for the docker container to use in the pod.
+                                                           :container-image-spec "twosigma/kitchen:latest"}
```
nitpick: we can name this as `container-image`

DaoWen replied:
Renamed to `:default-container-image` based on Steven's feedback. Adding "default" to the name should help make things more clear when we start allowing users to override the default image.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -588,9 +524,79 @@
+     :metadata {:name k8s-name
+                :labels {:app k8s-name
+                         :managed-by orchestrator-name}
+                :annotations {:waiter-service-id service-id}}
```
Nitpick: lexicographic order

DaoWen replied:
Sorted.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -588,9 +524,79 @@
+                                     :managed-by orchestrator-name}}
+            :template {:metadata {:labels {:app k8s-name
+                                           :managed-by orchestrator-name}
+                                  :annotations {:waiter-port-count (str ports)
```
Nitpick: lexicographic order

DaoWen replied:
Sorted.

DaoWen commented:
@shamsimam @sradack - I changed how the ReplicaSet templates are handled. The template is now inline in the Clojure code, and the factory function is specified in the config. Please review again when you have time.

sradack commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,653 @@
+                                        service-id->password-fn home-path)
+        ;; Randomize $PORT0 value to ensure clients can't hardcode it.
+        ;; Helps maintain compatibility with Marathon, where port assignment is dynamic.
+        port0 (-> (rand-int 100) (* 10) (+ pod-base-port))
```
It would be nice to have a consistent `ReplicaSet` definition for a given `service-id`.  The "randomness" could be seeded.

DaoWen replied:
`port0` is now generated based on the `hash` of the `service-id` string:

```clojure
port0 (-> service-id hash (mod 100) (* 10) (+ pod-base-port))
```

This makes the port number deterministic for a given service description, but still practically unpredictable.

sradack commented:
@shamsimam has some unaddressed comments on this PR as well.

DaoWen commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,646 @@
+         :healthy? (get-in pod [:status :containerStatuses 0 :ready] false)
+         :host (get-in pod [:status :podIP])
+         :id (pod->instance-id pod)
+         :log-directory (str "/home/" (get-in pod [:metadata :namespace]))
```
Oops! Sorted.

DaoWen commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,646 @@
+
+(defn- service-description->namespace
+  [service-description]
+  (get service-description "run-as-user"))
```
Fixed.

DaoWen commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,646 @@
+  (and (some? (get-in pod [:status :podIP]))
+       (nil? (get-in pod [:metadata :deletionTimestamp]))))
+
+(defn- get-service-instances
```
Renamed.

DaoWen commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -0,0 +1,646 @@
+       (nil? (get-in pod [:metadata :deletionTimestamp]))))
+
+(defn- get-service-instances
+  "Get all Waiter Service Instances associated with the given Waiter Service."
```
Docstring updated.

I like the idea of moving the failed-instance-tracking trigger, but currently the pod descriptions don't escape `get-service-instances!`. I feel like the complexity of letting the k8s-pod objects escape that scope negatively-offsets the value of moving the failed-instance-tracking logic.

DaoWen commented:
@shamsimam @sradack - Please review the changes new commits I pushed this morning:

* Addressed Sham's outstanding feedback (including nitpicks).
* Refactored `port0` value generation per Steven's feedback.
* Added the Kuberentes Pod random suffix length as an optional setting in the Waiter config.
* Refactored the authentication string `refresh-fn` to take its context as an argument (rather than being nullary), which fits better with our waiter factory-function pattern.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -534,9 +534,9 @@
+        ;; Make $PORT0 value pseudo-random to ensure clients can't hardcode it.
         ;; Helps maintain compatibility with Marathon, where port assignment is dynamic.
-        port0 (-> (rand-int 100) (* 10) (+ pod-base-port))
+        port0 (-> service-id hash (mod 100) (* 10) (+ pod-base-port))
```
Nitpick: Why do we want the port number to end in zero? wouldn't doing `(-> service-id hash (mod 1000) (+ pod-base-port))` increase the port range?

shamsimam replied:
I found it odd since don't have that restriction on Marathon services, but am okay with whatever we decide.

shamsimam commented on waiter/test/waiter/scheduler/kubernetes_test.clj:
```diff
@@ -875,4 +881,17 @@
+          (try
+            (with-redefs [start-auth-renewer #(reset! kill-task-fn (apply orig-start-auth-renewer %&))]
+              (is (instance? KubernetesScheduler (kubernetes-scheduler (assoc base-config :authentication {:refresh-delay-mins 1
+                                                                                                           :refresh-fn `test-auth-refresher
```
It'd be nice to settle on a convention for the entry name for the function symbol, e.g. `:fn`, in all the places in the config where we use this pattern. This entry can then be placed as the first entry of the config map (like we do for `kind` or `factory-fn`).

DaoWen replied:
The problem with `:fn` is that it becomes `fn` when destructuring, and we don't want to shadow a core symbol like that. (I tried that before with the old `:edn-params` option—it was really ugly.)

> This entry can then be placed as the first entry of the config map

Wait... You're saying that you *don't* want it sorted lexicographically?! 😮

DaoWen replied:
If we decide to rename all the fn-symbol entries to use a common keyword, that would need to go in a separate PR. If you feel strongly about that, you should probably open a new issue for that task.

shamsimam replied:
I think we should settle on a name and use it, `:fn` may not be that name due to the issue you mentioned.

I searched `config-full.edn` and didn't find another place where we have a non-`factory-fn`. Since, this is the first PR that's introducing this name, it makes sense for this PR to also introduce that standard.

DaoWen replied:
Renamed `:refresh-fn` to `:action-fn` per our offline discussion.

shamsimam commented on waiter/config-full.edn:
```diff
@@ -278,16 +278,17 @@
+                                                  ;; If omitted, then the :action-fn value is only called once at Waiter startup.
+                                                  :refresh-delay-mins 15}
+
+                                                  ;; Unary function which can be called periodically (or just once if :refresh-delay-mins is omitted)
```
Is this string intentionally placed here?

DaoWen replied:
Nope. Cut/paste error. Fixed now.

shamsimam commented on waiter/src/waiter/scheduler/kubernetes.clj:
```diff
@@ -596,11 +596,11 @@
   "Initialize the k8s-api-auth-str atom,
    and optionally start a chime to periodically refresh the value."
-  [{:keys [refresh-delay-mins refresh-fn] :as context}]
+  [{:keys [refresh-delay-mins action-fn] :as context}]
```
Nitpick: lexicographic order
